lib/src/endpoint_paginator.dart

```diff
@@ -0,0 +1,290 @@
+import 'package:nyxx/nyxx.dart';
+
+/// Controls the order in which entities from paginated endpoints are streamed.
+enum StreamOrder {
+  /// Emit the entities in order of most recent to oldest.
+  mostRecentFirst,
+
+  /// Emit the entities on order of oldest to most recent.
+  oldestFirst,
+}
+
+/// Wrap the paginated API call [fetchPage] into a stream.
+///
+/// Although this function supports bi-directional emitting of events using the
+/// [order] parameter, it can be used for API endpoints that only support
+/// pagination in one direction by hard-coding the [order] parameter to match
+/// the API order.
+Stream<T> _streamPaginatedEndpoint<T>(
+  Future<List<T>> Function({Snowflake? before, Snowflake? after, int? limit}) fetchPage, {
+  required Snowflake Function(T) extractId,
+  required Snowflake? before,
+  required Snowflake? after,
+  required int? pageSize,
+  required StreamOrder? order,
+}) async* {
+  // Both after and before:    oldest first
+  // Only after:               oldest first
+  // Only before:              most recent first
+  // Neither after nor before: oldest first
+  order ??= before != null && after == null ? StreamOrder.mostRecentFirst : StreamOrder.oldestFirst;
+  before ??= Snowflake.now();
+  after ??= Snowflake.zero;
+
+  var nextPageBefore = before;
+  var nextPageAfter = after;
+
+  while (true) {
+    // We choose the order of the pages by passing either before or after
+    // depending on the stream order.
+    final page = await switch (order) {
+      StreamOrder.mostRecentFirst => fetchPage(limit: pageSize, before: nextPageBefore),
+      StreamOrder.oldestFirst => fetchPage(limit: pageSize, after: nextPageAfter),
+    };
+
+    if (page.isEmpty) break;
```

Comment:
please use {} always. Maybe its not bad it that case but without {} code can get hard to read very fast

Reply:
Fixed!